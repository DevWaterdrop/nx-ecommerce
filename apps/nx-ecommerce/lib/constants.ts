export const LOVELY_ERROR = 'I HATE TYPESCRIPT+STRAPI+GRAPHQL';
export const MAXIMUM_IN_CART = 16;

export enum PRICE_FILTER {
  ZERO_TWENTY = 'zt',
  TWENTY_FORTY = 'tf',
  FORTY_EIGHTY = 'fe',
  EIGHTY_PLUS = 'ep',
}

export enum SORT_FILTER {
  PRICE_LOW_HIGH = 'plh',
  PRICE_HIGH_LOW = 'phl',
  NAME = 'name',
}

export const PRICES = [
  { value: PRICE_FILTER.ZERO_TWENTY, label: '0.00$ - 19.99$' },
  { value: PRICE_FILTER.TWENTY_FORTY, label: '20.00$ - 39.99$' },
  { value: PRICE_FILTER.FORTY_EIGHTY, label: '40.00$ - 79.99$' },
  { value: PRICE_FILTER.EIGHTY_PLUS, label: '80.00$+' },
];

export const SORT = [
  { value: SORT_FILTER.PRICE_LOW_HIGH, label: 'Price: low to high' },
  { value: SORT_FILTER.PRICE_HIGH_LOW, label: 'Price: high to low' },
  { value: SORT_FILTER.NAME, label: 'Name' },
];
