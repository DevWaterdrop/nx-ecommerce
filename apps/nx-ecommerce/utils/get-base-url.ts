export function getBaseURL(path = '') {
  return `${
    (process.env['NX_BASE_URL'] as string) || 'http://localhost:1337'
  }${path}`;
}
