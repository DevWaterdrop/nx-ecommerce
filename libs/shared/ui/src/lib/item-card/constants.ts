import { ItemCardProps } from './item-card';
import flowerJPEG from '../../props/flower.jpeg';
import roomJPEG from '../../props/room.jpeg';

const ITEM: Required<ItemCardProps['item']> = {
  price: 65,
  name: 'Lovely flower',
  id: '#',
  images: [flowerJPEG, roomJPEG],
  smallDescription: 'For you, for me, for us',
  tag: 'new',
};

export const DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleClick: () => console.log('clicked'),
};
