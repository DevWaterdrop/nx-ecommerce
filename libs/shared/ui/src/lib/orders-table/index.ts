import { OrdersTable, OrdersTableProps } from './orders-table';

export { OrdersTable };
export type { OrdersTableProps };
