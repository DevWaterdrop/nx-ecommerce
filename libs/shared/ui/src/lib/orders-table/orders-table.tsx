import { Product } from '@nx-ecommerce/shared/graphql/types';
import clsx from 'clsx';
import { Head } from './local-components/head';
import { Order } from './local-components/item';

export type Order = {
  id: string;
  amount: number;
  items: { id: string; amount: number; product: Product }[];
};

export interface OrdersTableProps {
  orders: Order[];
}

export const OrdersTable: React.FC<OrdersTableProps> = (props) => {
  const { orders } = props;

  return (
    <table className={clsx('min-w-full', 'divide-y border divide-gray-300')}>
      <Head />
      <tbody className="divide-y bg-white divide-gray-200">
        {orders.map((order) => (
          <Order order={order} key={order.id} />
        ))}
      </tbody>
    </table>
  );
};
