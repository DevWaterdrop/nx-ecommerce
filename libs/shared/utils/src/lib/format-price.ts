export function formatPrice(price: number) {
  return `${Math.max(price, 0).toFixed(2)}$`;
}
