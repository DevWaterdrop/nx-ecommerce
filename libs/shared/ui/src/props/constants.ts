import { ButtonProps } from '../lib/button';
import { CartItemProps } from '../lib/cart-item';
import { CartSidebarProps } from '../lib/cart-sidebar';
import { CategoryCardProps } from '../lib/category-card';
import { IconProps } from '../lib/icon/icon';
import { ItemCardProps } from '../lib/item-card';
import { NavButtonProps } from '../lib/nav-button';
import { OrdersTableProps } from '../lib/orders-table';
import { ProductGalleryProps } from '../lib/product-gallery';
import { ProductInformationProps } from '../lib/product-information';
import { ProductSidebarProps } from '../lib/product-sidebar';
import { SearchbarProps } from '../lib/searchbar';
import { SearchbarItemProps } from '../lib/searchbar-item';
import { SelectProps } from '../lib/select';
import flowerJPEG from './flower.jpeg';
import roomJPEG from './room.jpeg';

export const ITEM: ItemCardProps['item'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
};

export const ITEM_SECOND: ItemCardProps['item'] = {
  slug: 'roomy',
  price: 50,
  name: 'Roomy',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'Hmmm, nice vase',
  description: 'Vaaaase',
};

export const ITEM_CARD_DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleCartClick: () => console.log('cart clicked'),
  handleAmountCartClick: () => console.log('amount cart clicked'),
  handleFavoriteClick: () => console.log('favorite clicked'),
};

export const CATEGORY_CARD_DEFAULT_PROPS: CategoryCardProps = {
  category: {
    smallDescription:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
    image: {
      data: {
        attributes: {
          url: roomJPEG,
        },
      },
    },
    slug: '1',
  },
};

export const NAV_BUTTON_DEFAULT_PROPS: NavButtonProps = {
  type: 'heart',
};

export const ICON_DEFAULT_PROPS: IconProps = {
  icon: 'cart',
};

export const BUTTON_DEFAULT_PROPS: ButtonProps = {
  children: 'hello',
  tag: 'button',
  size: 'base',
  type: 'primary',
  round: 'full',
};

export const SEARCHBAR_ITEM: SearchbarProps['items'][number] = {
  id: '111',
  attributes: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
};

export const SEARCHBAR_ITEM_DEFAULT_PROPS: SearchbarItemProps = {
  item: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
  handleClick: () => void 1,
};

export const SELECT_ITEM = {
  label: 'Label1',
  value: 'value1',
};

export const SELECT_DEFAULT_PROPS: Required<
  Omit<SelectProps, 'placeholder' | 'multiple'>
> = {
  items: [SELECT_ITEM],
  initial: ['value1'],
  onChange: () => console.log('changed'),
};

export const PRODUCT: ProductSidebarProps['product'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
  information: [
    {
      title: 'Title1',
      content: 'content1',
    },
  ],
};

export const PRODUCT_SIDEBAR_DEFAULT_PROPS: ProductSidebarProps = {
  product: PRODUCT,
  handleCartClick: () => {
    console.log('cart');
  },
  handleRemoveCartClick: () => {
    console.log('remove cart');
  },
  handleFavoriteClick: () => {
    console.log('favorite');
  },
  cartValue: new Map([]),
  favoritesValue: new Set('lovely-flower'),
};

export const PRODUCT_INFORMATION_DEFAULT_PROPS: ProductInformationProps = {
  information: [...Array(10).keys()].map((_value, index) => ({
    title: `Title${index + 1}`,
    content: `content${index + 1}`,
  })),
};

export const PRODUCT_GALLERY_DEFAULT_PROPS: ProductGalleryProps = {
  images: [
    {
      attributes: {
        url: flowerJPEG,
      },
    },
    {
      attributes: {
        url: roomJPEG,
      },
    },
  ],
};

const PRODUCT_ORDER_TABLE: OrdersTableProps['orders'][number]['items'][number]['product'] =
  {
    name: 'Flower',
    slug: 'slug-flower',
    description: '',
    smallDescription: '',
    price: 50,
  };

export const ORDERS_TABLE_DEFAULT_PROPS: OrdersTableProps = {
  orders: [
    {
      id: 'id-1',
      amount: 50,
      items: [{ id: '1', amount: 1, product: PRODUCT_ORDER_TABLE }],
    },
    {
      id: 'id-2',
      amount: 100,
      items: [{ id: '2', amount: 2, product: PRODUCT_ORDER_TABLE }],
    },
  ],
};

export const CART_SIDEBAR_DEFAULT_PROPS: CartSidebarProps = {
  cartTotal: 100,
  isUser: true,
  handleCheckoutClick: async () => {
    console.log('checkout');
  },
  isLoading: false,
};

export const CART_ITEM_DEFAULT_PROPS: CartItemProps = {
  inCartAmount: 1,
  item: PRODUCT,
  handleCartMinusClick: () => {
    console.log('cart minus');
  },
  handleCartPlusClick: () => {
    console.log('cart plus');
  },
  handleRemoveFromCart: () => {
    console.log('cart remove');
  },
  toggleFavorite: () => {
    console.log('favorite');
  },
};
