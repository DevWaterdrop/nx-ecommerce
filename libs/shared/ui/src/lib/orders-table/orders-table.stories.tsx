import { Story, Meta } from '@storybook/react';
import { ORDERS_TABLE_DEFAULT_PROPS } from '../../props/constants';
import { OrdersTable, OrdersTableProps } from './orders-table';

export default {
  component: OrdersTable,
  title: 'Components/OrdersTable',
} as Meta;

const Template: Story<OrdersTableProps> = (args) => <OrdersTable {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  ...ORDERS_TABLE_DEFAULT_PROPS,
};
