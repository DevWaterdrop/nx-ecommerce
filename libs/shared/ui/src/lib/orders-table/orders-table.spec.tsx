import { render, RenderResult } from '@testing-library/react';
import '@testing-library/jest-dom';

import { OrdersTable } from './orders-table';
import { ORDERS_TABLE_DEFAULT_PROPS } from '../../props/constants';
import { formatPrice } from '@nx-ecommerce/shared/utils/format-price';

describe('OrdersTable', () => {
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(<OrdersTable {...ORDERS_TABLE_DEFAULT_PROPS} />);
  });

  it('should render successfully', () => {
    const { baseElement } = renderResult;
    expect(baseElement).toBeTruthy();
  });

  it('should have proper head structure', () => {
    const { baseElement } = renderResult;

    const tableHead = baseElement.querySelector('thead');
    expect(tableHead).toBeInTheDocument();

    const headRow = tableHead?.querySelector('tr');
    expect(headRow).toBeInTheDocument();

    const tableDataCells = headRow?.querySelectorAll('td');
    expect(tableDataCells?.length).toBe(2);

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const [products, total] = Array.from(tableDataCells!);

    expect(products.textContent).toBe('Products');
    expect(total.textContent).toBe('Total');
  });

  it('should have proper body structure', () => {
    const { baseElement } = renderResult;

    const { orders } = ORDERS_TABLE_DEFAULT_PROPS;

    const tableBody = baseElement.querySelector('tbody');
    expect(tableBody).toBeInTheDocument();

    orders.forEach((order, index) => {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const BodyRow = Array.from(tableBody!.querySelectorAll('tr'))[index];
      expect(BodyRow).toBeInTheDocument();

      const tableDataCells = BodyRow?.querySelectorAll('td');
      expect(tableDataCells?.length).toBe(2);

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const [products, amount] = Array.from(tableDataCells!);

      expect(products.querySelectorAll('li').length).toBe(order.items.length);
      expect(amount.textContent).toBe(formatPrice(order.amount));
    });
  });
});
