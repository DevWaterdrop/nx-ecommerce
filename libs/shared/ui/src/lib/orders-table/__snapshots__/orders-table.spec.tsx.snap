// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersTable should render successfully 1`] = `
<table
  className="min-w-full divide-y border divide-gray-300"
>
  <thead
    className="bg-gray-50"
  >
    <tr
      className="divide-x divide-gray-200"
    >
      <td
        className="py-3.5 pr-4 pl-4 font-semibold text-left text-sm sm:(pl-6)"
      >
        Products
      </td>
      <td
        className="py-3.5 pr-4 pl-4 font-semibold text-left text-sm sm:(pr-6)"
      >
        Total
      </td>
    </tr>
  </thead>
  <tbody
    className="divide-y bg-white divide-gray-200"
  >
    <tr
      className="divide-x divide-gray-200"
    >
      <td
        className="py-4 pr-4 pl-4 font-medium text-sm whitespace-nowrap sm:(pl-6)"
      >
        <ul
          className="flex flex-col gap-1"
        >
          <li
            className="py-2 flex gap-4 items-center border-b"
          >
            <span>
              1
            </span>
            <a
              className="text-sky-600 uppercase hover:(underline)"
              href="/i/slug-flower"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Flower
            </a>
          </li>
        </ul>
      </td>
      <td
        className="py-4 pr-4 pl-4 w-4 font-medium text-xl whitespace-nowrap sm:(pr-6)"
      >
        50.00$
      </td>
    </tr>
    <tr
      className="divide-x divide-gray-200"
    >
      <td
        className="py-4 pr-4 pl-4 font-medium text-sm whitespace-nowrap sm:(pl-6)"
      >
        <ul
          className="flex flex-col gap-1"
        >
          <li
            className="py-2 flex gap-4 items-center border-b"
          >
            <span>
              2
            </span>
            <a
              className="text-sky-600 uppercase hover:(underline)"
              href="/i/slug-flower"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Flower
            </a>
          </li>
        </ul>
      </td>
      <td
        className="py-4 pr-4 pl-4 w-4 font-medium text-xl whitespace-nowrap sm:(pr-6)"
      >
        100.00$
      </td>
    </tr>
  </tbody>
</table>
`;
